### PR TITLE
go/worker/executor: cache of last seen runtime transactions

### PR DESCRIPTION
**Pull request by ptrus**
Closes: #3274

### CODE REVIEW

File: go/worker/compute/executor/committee/node.go [hunk 1]
```diff
@@ -593,10 +607,28 @@ func (n *Node) QueueTx(call []byte) error {
 	if n.scheduler == nil || !n.scheduler.IsInitialized() {
 		return errNotReady
 	}
+
+	callHash := hash.NewFromBytes(call)
+	if _, b := n.lastScheduledCache.Peek(callHash); b {
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:612]: Why `Peek` instead of `Get`?

File: go/worker/compute/executor/committee/node.go [hunk 2]
```diff
@@ -432,11 +436,21 @@ func (n *Node) HandleEpochTransitionLocked(epoch *committee.EpochSnapshot) {
 	if !epoch.IsExecutorWorker() {
 		// Clear incoming queue if we are not a worker anymore.
 		n.scheduler.Clear()
+		n.lastScheduledCache.Clear()
 		incomingQueueSize.With(n.getMetricLabels()).Set(0)
 	}
 
 	switch {
 	case epoch.IsExecutorMember():
+		if state, ok := n.state.(StateWaitingForFinalize); ok {
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:445]: Hm on a related note, in `HandleNewBlockLocked` when in `StateWaitingForFinalize` state we currently always call `removeTxBatch` but we likely need to check that the round succeeded (header type is "normal") and that this was actually the round that we processed (e.g. by comparing the IO roots) as otherwise we can incorrectly remove a batch that was not actually finalized?

File: go/worker/compute/executor/committee/node.go [hunk 3]
```diff
@@ -1423,6 +1456,11 @@ func NewNode(
 		return nil, err
 	}
 
+	cache, err := lru.New(lru.Capacity(lastScheduledCacheSize, false))
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:1459]: If cache size is specified as zero this should probably disable the cache instead of making it unlimited (which sounds dangerous) which is what happens currently?

File: go/worker/compute/executor/committee/node.go (hunk 2 above)
Comment (ptrus) [reply go/worker/compute/executor/committee/node.go:445]: Yeah good point, i did notice that, but then only checked what happens on epoch transition, which is handled fine (we don't delete the batch since we transition the state before that already), didnt thought about other failures though. Will fix it

Comment (ptrus) [reply go/worker/compute/executor/committee/node.go:447]: done

File: go/worker/compute/executor/committee/node.go [hunk 4]
```diff
@@ -492,28 +510,46 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
...
-		// A new block means the round has been finalized.
-		n.logger.Info("considering the round finalized",
-			"round", blk.Header.Round,
-			"header_hash", blk.Header.EncodedHash(),
-		)
-		// Record time taken for successfully processing a batch.
-		batchProcessingTime.With(n.getMetricLabels()).Observe(time.Since(state.batchStartTime).Seconds())
+		func() {
+			defer n.transitionLocked(StateWaitingForBatch{})
+			// A new block means the round has been finalized.
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:520]: Empty line above.

File: go/worker/compute/executor/committee/node.go [hunk 5]
```diff
@@ -1412,7 +1467,10 @@ func NewNode(
 	roleProvider registration.RoleProvider,
 	scheduleCheckTxEnabled bool,
 	scheduleMaxQueueSize uint64,
+	lastScheduledCacheSize uint64,
 ) (*Node, error) {
+	var err error
```
Comment (kostko) [inline go/worker/compute/executor/committee/node.go:1472]: Why is this needed, isn't `err` defined when `NewRuntimeHostNode` is called below?

File: go/worker/compute/executor/committee/node.go (hunk 2 above)
Comment (kostko) [reply go/worker/compute/executor/committee/node.go:447]: Looks good, thanks!

**Review by kostko — APPROVED** [review]

File: go/worker/compute/executor/committee/node.go (hunk 5 above)
Comment (ptrus) [reply go/worker/compute/executor/committee/node.go]: You are right, thx :+1:
